When daemon is in startup process, could not start container

moypray opened the pull request:
Description:
 When docker is in startup process and containerd sends an "process exit" event to docker.
If the container config '--restart=always', restartmanager will start this
container very soon.

But some initialization is not done, e.g. `daemon.netController`, when
visit, docker would panic.



<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Fix https://github.com/moby/moby/issues/33584

**- How I did it**

Start containerd event monitor after daemon finished restoring containers from disk.

**- How to verify it**

1. startup lots of containers and stop docker service.
2. kill some container process.
3. start up the docker service.

Will not see any `process exit` events received from containerd before docker restore containers.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Looks like there's a golint issue;

```
09:39:27 Errors from golint:
09:39:27 daemon/daemon.go:766:1: comment on exported method Daemon.WaitToWork should be of the form "WaitToWork ..."
09:39:27 
09:39:27 Please fix the above errors. You can test via "golint" and commit the result.
09:39:27 
```

moypray commented:
Thx,  there is another compiling issue, I will fix them soon.

moypray committed:
When daemon is in startup process, could not start container

Description:
 When docker is in startup process and containerd sends an "process exit" event to docker.
 If the container config '--restart=always', restartmanager will start this container very soon.

 But some initialization is not done, e.g. `daemon.netController`,when visit, docker would panic.

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

mlaventure commented:
I don't think this fixes the issue unfortunately,  `daemon.containerd.Restore()` queries `containerd` about the last events that were generated since the last one docker is known to have processed. If there's an `ExitEvent` for a container with `--restart=always` in there, `daemon.StateChanged()` will still be called and it will schedule a go routine to restart the container. Meaning, it's basically up to scheduling whether that gets executed before the network controller is initialized.

I've been integrating `containerd-1.0` in docker, and the way I've done it could work in here too: I've made it so that `libcontainerd.Restore()` never calls `daemon.StateChanged()`, instead it returns whether a container is alive or not; this leave the `daemon.Restore()` decides on whether a restart is necessary or not (that code being actually already present).

/cc @cpuguy83 @tonistiigi

moypray commented:
@mlaventure , Thanks for your comments.

After checking the code, there are 2 ways to call `StateChange`,  One is `daemon.containerd.Restore` as you mentioned;  another is in `handleEventStream`. I just notice the second one....pity.   :(=   :(-

Your suggestion just fix the first conditions.(`Restore` does not call `StateChange`)

And I have an earliest fix is that add an `chan struct{}` to docker daemon, when docker daemon restore done, will close it. and in `daemon.StateChanged` starting container goroutine will wait for/read this chan.

I think this could fix this issue, but only solve this issue.  think about this as below:
Docker could receive containerd event during restore process and maybe containerd send an `ExitEvent` of an container which is not loaded by docker.  this event could be handled by error:
https://github.com/moby/moby/blob/master/daemon/monitor.go#L32

Will `handleEventStream` call `StateChange` later in your plan?

I will rework my PR after we have an better solution.

thaJeztah commented:
ping @mlaventure ^^

mlaventure commented:
@moypray true, `StateChanged` should be synchronized. The channel would work, but I think I prefer the change you made it to avoid a useless select at every call of `StateChanged` afterwards.

Do you feel comfortable making the needed changes to `libcontainerd` to handle the restore case too?

mlaventure commented:
Also, I'm not sure why `TestDockerNetworkHostModeUngracefulDaemonRestart` is failing with the current changes

moypray commented:
hi @mlaventure   I have reworked my PR. I add a chan to deamon to sync the status and close it after restore the containers.
Wait the chan in `StateChanged`, no matter `EventMonitor` or `containerd.Restore` called StateChanged, docker won't panic.

Wait for CI.

mlaventure commented:
I would have preferred avoiding having to check a channel every single time a restart is required, but I guess it required a more in depth refactoring.

/cc @tonistiigi wdyt? Is this a good enough stop gap?

tonistiigi commented:
@mlaventure I guess this is better than current.

mlaventure commented:
Alright, let's take it.

mlaventure reviewed (approved):
LGTM
